schematool/db/_hive.py

```diff
@@ -0,0 +1,207 @@
+
+    @classmethod
+    def get_commit_history(cls):
+        # Omit the placeholder row
+        return cls.execute('SELECT * FROM %s WHERE alter_hash != \'%s\'' %
+            (cls.full_table_name, cls.DUMMY_ALTER_REF))
+
+    @classmethod
+    def get_applied_alters(cls):
+        # Omit the placeholder row
+        results = cls.execute('SELECT alter_hash FROM %s WHERE alter_hash != \'%s\'' %
+            (cls.full_table_name, cls.DUMMY_ALTER_REF))
+        alters_hashes = [result[0] for result in results]
+        return alters_hashes
+
+    @classmethod
+    def append_commit(cls, ref):
+        return cls.execute(cls.get_append_commit_query(ref))
+
+    @classmethod
+    def get_append_commit_query(cls, ref):
+        # Use a monotonically increasing value for the id field, since Hive does not have an
+        # auto-increment mechanism
+        id_value = int(time())
+        ran_on = datetime.strftime(datetime.now(), '%Y-%m-%d %H:%M:%S')
+        return ("""INSERT INTO TABLE %s
+                   SELECT %s AS id, '%s' AS alter_hash, '%s' AS ran_on FROM %s
+                   WHERE alter_hash = '%s' LIMIT 1""" %
+            (cls.full_table_name, id_value, ref, ran_on, cls.full_table_name, cls.DUMMY_ALTER_REF))
+
+    @classmethod
+    def remove_commit(cls, ref):
+        return cls.execute(cls.get_remove_commit_query(ref))
+
+    @classmethod
+    def get_remove_commit_query(cls, ref):
+        return ("""INSERT OVERWRITE TABLE %s SELECT * FROM %s
+                   WHERE alter_hash != '%s'""" % (cls.full_table_name, cls.full_table_name, ref))
+
+    @classmethod
+    def create_history(cls):
+        create_table_result = cls.execute("""CREATE TABLE IF NOT EXISTS %s (
+                                                id int,
+                                                alter_hash string,
+                                                ran_on timestamp
+                                            ) ROW FORMAT DELIMITED FIELDS TERMINATED BY '\t'
+                                            STORED AS TEXTFILE""" % cls.full_table_name)
+
+        # Check whether table has been bootstrapped with the intial placeholder entry
+        results = cls.execute('SELECT * FROM %s WHERE alter_hash = \'%s\'' %
+            (cls.full_table_name, cls.DUMMY_ALTER_REF))
+
+        if not results:
+            # Create a tab-separated file that will be used to seed the history table, and attempt
+            # to load it via Hive.
+            #
+            # Note: Due to Hive's limited native support for certain SQL operations, like
+            # 'INSERT INTO ... VALUES', 'UPDATE', and 'DELETE FROM', we must seed this initial
+            # "pivot" row into the history table so that we can simulate standard insert and delete
+            # operations using 'INSERT INTO ... SELECT' syntax. This is a bit roundabout, but seems
+            # to be most compatible with our particular Hive/Hadoop installation.
```

Comment:
Good point - this decision was informed by our particular setup, but there is nothing AppNexus-specific here.

With more recent versions of Hive, the SQL constructs I mentioned (INSERT...VALUES/UPDATE/DELETE) are actually supported, but with some caveats: https://cwiki.apache.org/confluence/display/Hive/Hive+Transactions (see the Limitations section)

In our case, we will soon be moving to a newer version of Hive with transactional support, but I don't believe we have any intention of enabling it. In writing this module, I decided to go with the more compatible (though slightly more complex) implementation, since it would be more likely to work with a given Hadoop installation. I can update the comments accordingly - though do you think there is something here that should be rethought?